File: 3699-number-of-zigzag-arrays-i/3699-number-of-zigzag-arrays-i.py
class Solution:
    def zigZagArrays(self, n: int, l: int, r: int) -> int:
        MOD = 1000000007
        m = r - l + 1
        dp = [1] * m
        for _ in range(2, n + 1):
            dp.reverse()
            pref = 0
            for i in range(m):
                old = dp[i]
                dp[i] = pref
                pref = (pref + old) % MOD
        ans = sum(dp) % MOD
        return (ans * 2) % MOD

File: 3699-number-of-zigzag-arrays-i/3699-number-of-zigzag-arrays-i.cpp
class Solution {
public:
    static constexpr int MOD = 1000000007;
    int zigZagArrays(int n, int l, int r) {
        int m = r - l + 1;
        vector<int> dp(m, 1);
        for (int len = 2; len <= n; len++) {
            reverse(dp.begin(), dp.end());
            long long pref = 0;
            for (int i = 0; i < m; i++) {
                int old = dp[i];
                dp[i] = pref;
                pref = (pref + old) % MOD;
            }
        }
        long long ans = 0;
        for (int x : dp) {
            ans = (ans + x) % MOD;
        }
        return (ans * 2) % MOD;
    }
};